velox/dwio/dwrf/common/FileMetadata.cpp

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/dwio/dwrf/common/FileMetadata.h"
+#include "velox/dwio/dwrf/common/Common.h"
+
+namespace facebook::velox::dwrf {
+
+namespace detail {
+
+template <typename ProtoType, typename FooterType>
+static std::vector<std::unique_ptr<TypeProto>> convertToTypeProto(
+    FooterType& footer) {
+  std::vector<std::unique_ptr<TypeProto>> res;
```

Comment:
Would be nice if we can remove this vector allocation & copy as well.  Something like a wrapper over `google::protobuf::RepeatedPtrField<FormattedType>`.

Reply:
I could not figure out a way to wrap the `RepeatedPtrField` as it cannot be subclassed.
In my latest change, I templated `Footer` to route all the access to attributes to the underlying proto footer. In this case, Footer holds no data other than the proto footer pointer, hence no copy. I also remove interfaces that returns `RepeatedPtrField` and `RepeatedField`.

Reply:
You could make a composition+proxy instead of inheritance.  But since the use case for that is rare, we can omit these for now.

Reply:
A change here is that the return type of templated attributes becomes `unique_ptr`, e.g., `footer->types(0)` returns `unique_ptr<DataType>`. The reasons are:
1. Polymorphism is needed by `DataType`, so we either return reference or pointer
2. `Footer` doesn't hold any value of `DataType`, so no way to return by reference

Reply:
We may want to avoid that heap allocation even in the wrapper (it's much worse than virtual dispatch cost).  Can you make a wrapper like this?

```
struct DataType {
  enum { kDwrf, kOrc} format_;
  void* impl_;
}
```

Or even merge (bitwise or) `format_` with `impl_` if we can make sure `impl_` is always aligned to 4 bytes.

Reply:
Like this?
```
struct DataType {
  enum { kDwrf, kOrc} format_;
  void* impl_;

  TypeKind subtypesSize() const {
    if (format_ == kDwrf) {
        return reinterpret_cast<proto::Type*>(impl_)->subtypes_size();
    } else {
        return reinterpret_cast<proto::orc::Type*>(impl_)->subtypes_size();
    }
  }
};
```

Reply:
Yes, you can add some template magic to reduce the boilerplate.